cores/esp8266/uart.c

```diff
@@ -614,47 +669,39 @@ uart_swap(uart_t* uart, int tx_pin)
     switch(uart->uart_nr) 
     {
     case UART0:
-        if(((uart->tx_pin == 1 || uart->tx_pin == 2) && uart->tx_enabled) || (uart->rx_pin == 3 && uart->rx_enabled)) 
-        {
-            if(uart->tx_enabled) //TX
-            {
-                pinMode(uart->tx_pin, INPUT);
+        if(uart->tx_enabled) { //TX
+            uart_disable_pin(uart->tx_pin);
+        }
+        if(uart->rx_enabled) { //RX
+            uart_disable_pin(uart->rx_pin);
+        }
+
+        if(((uart->tx_pin == 1 || uart->tx_pin == 2) && uart->tx_enabled)
+                || (uart->rx_pin == 3 && uart->rx_enabled)) {
+            if(uart->tx_enabled) { //TX
                 uart->tx_pin = 15;
             }
-            if(uart->rx_enabled) //RX
-            {
-                pinMode(uart->rx_pin, INPUT);
+            if(uart->rx_enabled) { //RX
                 uart->rx_pin = 13;
             }
-            if(uart->tx_enabled) 
-                pinMode(uart->tx_pin, FUNCTION_4);    //TX
-
-            if(uart->rx_enabled)
-                pinMode(uart->rx_pin, FUNCTION_4);    //RX
-            
             IOSWAP |= (1 << IOSWAPU0);
-        } 
-        else 
-        {
-            if(uart->tx_enabled) //TX
-            {
-                pinMode(uart->tx_pin, INPUT);
+        } else {
+            if(uart->tx_enabled) { //TX
                 uart->tx_pin = (tx_pin == 2)?2:1;
             }
-            if(uart->rx_enabled) //RX
-            {
-                pinMode(uart->rx_pin, INPUT);
+            if(uart->rx_enabled) { //RX
                 uart->rx_pin = 3;
             }
-            if(uart->tx_enabled)
-                pinMode(uart->tx_pin, (tx_pin == 2)?FUNCTION_4:SPECIAL);    //TX
-
-            if(uart->rx_enabled) 
-                pinMode(3, SPECIAL);    //RX
-
             IOSWAP &= ~(1 << IOSWAPU0);
         }
 
+        if(uart->tx_enabled) { //TX
+            uart0_enable_tx_pin(uart->tx_pin);
```

Comment:
In the previous code, if pin 2, then FUNCTION_4, else SPECIAL.
In the current code, if pin 2 ||15, then FUNCTION_4, else SPECIAL.
The mode for pin 15 is different. This looks fishy to me.

Reply:
This was fixed in the last commit where the uart refactoring was removed.  All register settings are now the same as in master branch.